qualcomm.go

```diff
@@ -1,6 +1,8 @@
 package main
 
 const (
+	qualcommHdrLen = 6
```

Comment:
Yes, the original code was fairly opaque in terms if which specification it was complying with, hence why I've started to sprinkle code comments referring to sections of the HPAV specification.

The main difference between HPAV 1.0 and 1.1 packets is that the latter have a fragmentation and sequence information, hence the slightly larger header.

For example, for HPAV 1.0:
```c
typedef struct __packed qualcomm_hdr {
    uint8_t MMV;
    uint16_t MMTYPE;
    uint8_t OUI [ETHER_ADDR_LEN >> 1]; 
}
```
whereas for HPAV 1.1:
```c
typedef struct __packed qualcomm_fmi {
    uint8_t MMV;
    uint16_t MMTYPE;
    uint8_t FMSN;
    uint8_t FMID;
    uint8_t OUI [ETHER_ADDR_LEN >> 1]; 
}
```
I indeed agree that we need to do more thorough checking of the responses. At present, only the MME type is checked. The reference implementation even checks the ethertype (because you never know what random data will appear on a socket when you are working at such a low level).

This is all on my to-do list.

Reply:
Just a quick follow-up reply to myself, because this had me a bit confused: the open-plc-utils reference implementation checks the EtherType of responses in userspace (mme/Unwantedmessage.c):
```c
    if (ntohs (homeplug->ethernet.MTYPE) != ETH_P_HPAV)
    {   
#if defined (__WHYNOT__)
        error (0, 0, "Wrong Ethernet Frame Type: Received %04X while waiting for %04X", ntohs (homeplug->ethernet.MTYPE), ETH_P_HPAV);
#endif
        return (-1);
    }
```
It's a bit of a mystery why they do this, because an strace of `int6kstat` confirms that they are passing the EtherType in the `sll_protocol` to the socket bind, which would automatically filter non-HPAV frames:
```
bind(3, {sa_family=AF_PACKET, sll_protocol=htons(0x88e1 /* ETH_P_??? */), sll_ifindex=if_nametoindex("wlp0s20f3"), sll_hatype=ARPHRD_ETHER, sll_pkttype=PACKET_HOST, sll_halen=6, ...) = 0
```
The origin of that is in the `struct channel` (ether/channel.c) which is passed to `openchannel` (ether/openchannel.c), and this seems to be the method used by the majority of the tools in the open-plc-utils collection.

I don't think we need to do any additional checks for the EtherType of received frames in homeplug_exporter, because we are also already specifying that we only want to listen to HPAV frames, which will filter out all frames with EtherType other than 0x88e1:
```go
conn, err := packet.Listen(iface, packet.Raw, etherType, nil)
```

Reply:
Thanks for the info!